hack/helm/release/internal/helm/helm.go

```diff
@@ -0,0 +1,555 @@
+// read the Chart.yaml file, Unmarshaling it into a chart struct, ignoring
+// any Helm charts with names from the excludes list, and return a slice of charts
+// and any errors.
+func readCharts(chartsDir string, excludes []string) ([]chart, error) {
+	cs, err := filepath.Glob(filepath.Join(chartsDir, chartYamlGlob))
+	if err != nil {
+		return nil, fmt.Errorf("while searching for files matching pattern (%s): %w", chartYamlGlob, err)
+	}
+	// also check if the current directory given is a chart for the case of providing
+	// the releaser a full path to a single chart to process.
+	currentDirectoryCharts, err := filepath.Glob(filepath.Join(chartsDir, "Chart.yaml"))
+	if err != nil {
+		return nil, fmt.Errorf("while searching for files matching pattern (%s): %w", "Chart.yaml", err)
+	}
+	var charts []chart
+	for _, fullChartPath := range append(cs, currentDirectoryCharts...) {
+		fileBytes, err := os.ReadFile(fullChartPath)
+		if err != nil {
+			return nil, fmt.Errorf("while reading (%s): %w", fullChartPath, err)
+		}
+		var ch chart
+		if err = yaml.Unmarshal(fileBytes, &ch); err != nil {
+			return nil, fmt.Errorf("while unmarshaling (%s) to chart: %w", fullChartPath, err)
+		}
+		if contains(ch.Name, excludes) {
+			log.Printf("Excluding (%s) as it is in the excludes list", ch.Name)
+			continue
+		}
+		ch.fullPath = filepath.Dir(fullChartPath)
+		charts = append(charts, ch)
+	}
+
+	return charts, nil
+}
+
+// contains is a helper to determine if a given string is in a slice of strings.
+func contains(name string, names []string) bool {
+	for _, n := range names {
+		if n == name {
+			return true
+		}
+	}
+	return false
+}
+
+// uploadChartsConfig is the configuration for uploading Helm charts
+// and updating the GCS bucket Helm index file.
+type uploadChartsConfig struct {
+	releaseConf ReleaseConfig
+	noDeps      charts
+	withDeps    charts
+}
+
+// uploadChartsAndUpdateIndex will perform the following:
+// 1. Upload Helm charts with no dependencies to GCS bucket.
+// 2. Potentially update GCS Bucket Helm index.
+// 3. Potentially uploade Helm charts with dependencies to GCS bucket.
+// 4. Potentially update GCS Bucket Helm index a second time.
+func uploadChartsAndUpdateIndex(conf uploadChartsConfig) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+	tempDir, err := os.MkdirTemp(os.TempDir(), "charts")
+	if err != nil {
+		return fmt.Errorf("while creating temporary directory for charts: %w", err)
+	}
+	defer os.RemoveAll(tempDir)
+	// upload charts without dependencies
+	if err := uploadCharts(ctx, tempDir, conf.noDeps, conf.releaseConf); err != nil {
+		return err
+	}
+	if err := updateIndex(ctx, tempDir, conf); err != nil {
+		return err
+	}
+	if conf.releaseConf.DryRun {
+		log.Printf("Not processing charts with dependencies (%v) as the Helm index isn't being updated, or dry-run is set", conf.withDeps)
+		return nil
+	}
```

Comment:
Why do we shortcut in the middle of `uploadChartsAndUpdateIndex` to not process charts with dependencies? IIUC the index of the charts with deps will be invalid because we didn't update the index of the parent charts. But I wonder if it is really that bad? With dry-run we want to test as much as possible of the process and we upload anything.

Reply:
Ok, I guess that makes sense.  It's been removed.